### PR TITLE
Add calico/node configuration guide.

### LINKED ISSUE #284: Add "configuring calico/node" guide
calico-cni, calicoctl, and k8s-policy have a configuration guide.

Calico-node could use one as well.

### PR DESCRIPTION
**Pull request by djosborne**
Fixes #284

### CODE REVIEW

File: master/reference/node/configuration.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+---
+
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment        | Description                                | Schema | Default |
+|----------------------------------------------------------------|--------|---|--|
+| NODENAME | Node ID. | string | |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: Node _name_ in the Calico API.

File: master/reference/node/configuration.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+| HOSTNAME | Hostname. | string | |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: Hostname should be marked as deprecated.

File: master/reference/node/configuration.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+| NO_DEFAULT_POOLS | Prevents calico from creating a default pool if one does not exist. | boolean | false |
+| IP | IP address which calico will bind to. | ip address | 10.0.0.1 |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: calico should be capitalized across the board.

File: master/reference/node/configuration.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+| IP6 | IPv6 address which calico will bind to. | ipv6 address | fed80::1 |
+| AS | AS number for this node. | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. |can be `kubernetes` or `etcd` | etcd |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: scratch the "can be" part.  This can just be a comma separated list.

File: master/reference/node/configuration.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+| WAIT_FOR_DATASTORE | Prevent calico/node startup from initiating etcd connection until an etcd connection is made. | boolean | false |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: > from initiating etcd connection until an etcd connection is made

is a bit confusing.  Maybe "wait for connection to datastore before starting"

File: master/reference/node/configuration.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+| CALICO_LIBNETWORK_ENABLED | Enables running the docker-libnetwork plugin directly in the calico/node container. | boolean | true |
+| CALICO_NETWORKING_BACKEND | Describes which BGP networking backend to use| can be `gobgp`, `bird`, `none` | bird |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: Should also be a list.  We should scratch the backticks because it makes formatting terrible.

File: master/reference/node/configuration.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
+---
+title: Configuring calico/node
+---
+
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment        | Description                                | Schema | Default |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: Might be nice to make these consistent with the titles used for the calicoctl resources.

```
Field	Description	Accepted Values	Schema	Default
```

File: master/reference/node/configuration.md (hunk 5 above)
Comment (robbrockbank) [inline master/reference/node/configuration.md]: also, if connection fails then the container exits.

File: master/reference/node/configuration.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment | Description                              | Schema | Default |
+| ------------- | -------- | ------ | ----- |
+| NODENAME | Node ID. | string | |
+| NO_DEFAULT_POOLS | Prevents calico from creating a default pool if one does not exist. | boolean | false |
+| HOSTNAME | Hostname. | string | |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: Deprecated

File: master/reference/node/configuration.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment | Description                              | Schema | Default |
+| ------------- | -------- | ------ | ----- |
+| NODENAME | Node ID. | string | |
+| NO_DEFAULT_POOLS | Prevents calico from creating a default pool if one does not exist. | boolean | false |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: I believe the default is an empty string :-(

Any non-empty value is taken to be true.

File: master/reference/node/configuration.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+| HOSTNAME | Hostname. | string | |
+| IP | IP address which calico will bind to. | ip address | 10.0.0.1 |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: If an IP address is specified here, the address is written in to the node resource configuration, and will override any value previously configured.

If no IP address is specified, and an address has not yet been configured in the node resource, we will auto-detect an IPv4 address and configure the node resource with that address.  This means, if the node restarts and no --IP option is specified then the previously auto detected value will be used.

If a value of "autodetect" is specified, the IP address will be auto detected every time the node starts.  This will overwrite any value configured in the node resource.

File: master/reference/node/configuration.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+| IP6 | IPv6 address which calico will bind to. | ipv6 address | fed80::1 |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: If an IP address is specified here, the address is written in to the node resource configuration, and will override any value previously configured.

If no IP address is specified, and an address has not yet been configured in the node resource, we will not enable IPv6 routing.

If no IP address is specified and an IPv6 address has been previously configured in the node resource, we will use that address for IPv6 routing.

File: master/reference/node/configuration.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+| AS | AS number for this node. | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. |can be `kubernetes` or `etcd` | etcd |
+| WAIT_FOR_DATASTORE | Prevent calico/node startup from initiating etcd connection until an etcd connection is made. | boolean | false |
+| CALICO_LIBNETWORK_ENABLED | Enables running the docker-libnetwork plugin directly in the calico/node container. | boolean | true |
+| CALICO_NETWORKING_BACKEND | Describes which BGP networking backend to use| can be `gobgp`, `bird`, `none` | bird |
+| **etcd configuration** | | | |
+| ETCD_ENDPOINTS    | A comma separated list of etcd endpoints (optional, defaults to http://127.0.0.1:2379) | http://etcd1:2379 | |
+| ETCD_KEY_FILE     | Path to the etcd key file (optional)       | /etc/calico/key.pem | |
+| ETCD_CERT_FILE    | Path to the etcd client cert (optional)    | /etc/calico/cert.pem | |
+| ETCD_CA_CERT_FILE | Path to the etcd CA file (optional)        | /etc/calico/ca.pem | |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: There is also an interface prefix environment.  Chat to @gunjan5 about that.

File: master/reference/node/configuration.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+| ------------- | -------- | ------ | ----- |
+| NODENAME | Node ID. | string | |
+| NO_DEFAULT_POOLS | Prevents calico from creating a default pool if one does not exist. | boolean | false |
+| HOSTNAME | Hostname. | string | |
+| IP | IP address which calico will bind to. | ip address | 10.0.0.1 |
+| IP6 | IPv6 address which calico will bind to. | ipv6 address | fed80::1 |
+| AS | AS number for this node. | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. |can be `kubernetes` or `etcd` | etcd |
+| WAIT_FOR_DATASTORE | Prevent calico/node startup from initiating etcd connection until an etcd connection is made. | boolean | false |
+| CALICO_LIBNETWORK_ENABLED | Enables running the docker-libnetwork plugin directly in the calico/node container. | boolean | true |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: We should also discuss the volume mounts that are required for this container.   Although I'd be happy with this being a separate PR :-)

File: master/reference/node/configuration.md (hunk 12 above)
Comment (gunjan5) [inline master/reference/node/configuration.md]: See: https://github.com/projectcalico/calico-containers/pull/1390 and https://github.com/projectcalico/calico/pull/334/files

File: master/reference/node/configuration.md (hunk 1 above)
Comment (djosborne) [inline master/reference/node/configuration.md]: Huh?

Comment (caseydavenport) [inline master/reference/node/configuration.md]: I don't think we should be calling it the ID.

We could say "The name of the node object to use in the Calico API.  Will be created if it does not exist." which is more accurate.

File: master/reference/node/configuration.md [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+
+## Environment Variables
+
+| Environment | Description                              | Schema | Default |
+| ------------- | -------- | ------ | ----- |
+| NODENAME | Node ID. | string | |
+| NO_DEFAULT_POOLS | Prevents Calico from creating a default pool if one does not exist. | string | |
+| HOSTNAME | (Deprecated) The Hostname of this host. This is used as a unique identifier for the node. | string | |
+| IP | The IP address for Calico to bind to. When specified, the address is saved in the [node resource configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/node) for this host, overriding any previously configured value. When omitted, if an address **has** been saved in the node resource, then that value will be used. When omitted, if an address **has not** yet been configured in the node resource, the node will auto-detect an IPv4 address and configure the node resource with that address. This autodetection can be forced (even if a value has already been set in the node resource) by setting `$IP4` to "autodetect". Doing so will overwrite any value configured in the node resource. | ip address | 10.0.0.1 |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: 10.0.0.1 is not the default.

Same for IPv6

File: master/reference/node/configuration.md [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+| IP6 | IPv6 address which Calico will bind to. If specified, the address is written in to the node resource configuration, and will override any value previously configured. If an `$IP6` is specified, **and** an address has not yet been configured in the node resource, we will not enable IPv6 routing. If an `$IP6` is specified but an IPv6 address has been previously configured in the node resource, `$IP6` is used. | ipv6 address | fed80::1 |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: I'd lose the $ - I don't think they are necessary.

File: master/reference/node/configuration.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment | Description                              | Schema | Default |
+| ------------- | -------- | ------ | ----- |
+| NODENAME | A unique identifier for this host. | string | |
+| NO_DEFAULT_POOLS | Prevents Calico from creating a default pool if one does not exist. | string | |
+| HOSTNAME | (Deprecated) The Hostname of this host. This is used as a unique identifier for the node. When omitted, NODENAME is used. | string | |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: Not quite.  NODENAME is used preferentially.

Comment (robbrockbank) [inline master/reference/node/configuration.md]: Basically we still honor this environment, but it's lower priority.

File: master/reference/node/configuration.md [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+---
+
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment | Description                              | Schema | Default |
+| ------------- | -------- | ------ | ----- |
+| NODENAME | A unique identifier for this host. | string | |
```
Comment (robbrockbank) [inline master/reference/node/configuration.md]: When both the NODENAME and HOSTNAME environments are omitted it defaults to the hostname.

**Comment by robbrockbank** [comment]
LGTM.   I still think it worth adding a comment that the NODENAME defaults to use the system hostname if not specified... but then we also recommend that the user always supply it to ensure hostname changes don't muck up configuration.   Or maybe we should say all of that :-)

File: _data/master/navbars/reference.yml [hunk 18]
```diff
@@ -77,6 +77,10 @@ toc:
         path: /reference/calicoctl/setup/etcdv2
       - title: kubernetes api datastore
         path: /reference/calicoctl/setup/kubernetes
+- title: Calico Node
```
Comment (caseydavenport) [inline _data/master/navbars/reference.yml]: This should be `calico/node` since that's the name of the component.

File: master/reference/node/configuration.md [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+| CALICO_LIBNETWORK_IFPREFIX | Interface prefix to use for the network interface within the Docker containers that have been networked by the Calico driver. | string | cali |
+| CALICO_NETWORKING_BACKEND | Describes which BGP networking backend to use| gobgp, bird, none | bird |
+| **etcd configuration** | | | |
+| ETCD_ENDPOINTS    | A comma separated list of etcd endpoints (optional, defaults to http://127.0.0.1:2379) | http://etcd1:2379 | |
+| ETCD_KEY_FILE     | Path to the etcd key file (optional)       | /etc/calico/key.pem | |
+| ETCD_CERT_FILE    | Path to the etcd client cert (optional)    | /etc/calico/cert.pem | |
+| ETCD_CA_CERT_FILE | Path to the etcd CA file (optional)        | /etc/calico/ca.pem | |
+
+The ETCD_CA_FILE, ETCD_CERT_FILE, and ETCD_KEY_FILE
+environment variables are required when using Etcd with SSL/TLS.  The values
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: Etcd should be etcd

File: master/reference/node/configuration.md [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+| Environment | Description                              | Schema | Default |
+| ------------- | -------- | ------ | ----- |
+| NODENAME | A unique identifier for this host. | string | |
+| NO_DEFAULT_POOLS | Prevents Calico from creating a default pool if one does not exist. | string | |
+| HOSTNAME [Deprecated] | The Hostname of this host. This is used as a unique identifier for the node. This value is overridden by NODENAME. | string | |
+| IP | The IP address for Calico to bind to. When specified, the address is saved in the [node resource configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/node) for this host, overriding any previously configured value. When omitted, if an address **has** been saved in the node resource, then that value will be used. When omitted, if an address **has not** yet been configured in the node resource, the node will auto-detect an IPv4 address and configure the node resource with that address. This autodetection can be forced (even if a value has already been set in the node resource) by setting IP to "autodetect". Doing so will overwrite any value configured in the node resource. | ip address | |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: This is all useful information that users need to know... but it feels like a bit much for putting into a table.

Is there a better way we can format this?

Comment (caseydavenport) [inline master/reference/node/configuration.md]: We should also word this in terms of the Calico API, like we do for AS below.

"The IPv4 address to assign this Node."

The actual "what is this value used for?" should be part of the API documentation on the Node resource.

Comment (robbrockbank) [inline master/reference/node/configuration.md]: @caseydavenport we spoke about this, one thing I suggested was a table of all of the environments with a single line summary of what it is, followed by a section of heading/description -- so that way we could pad out the details without filling up a table.

We figured having all the information in a table would be a good increment - it wasn't clear initially if we'd need it.

File: _data/v2_0/navbars/reference.yml [hunk 21]
```diff
@@ -72,11 +72,15 @@ toc:
   - title: Setup
     section:
       - title: Overview
-        path: /reference/calicoctl/setup/
+        path: /reference/calicoctl/setup
       - title: etcdv2 datastore
         path: /reference/calicoctl/setup/etcdv2
       - title: kubernetes api datastore
         path: /reference/calicoctl/setup/kubernetes
+- title: Calico Node
```
Comment (caseydavenport) [inline _data/v2_0/navbars/reference.yml]: This needs to be changed like you did above.

File: master/reference/node/configuration.md [hunk 22]
```diff
@@ -0,0 +1,32 @@
+---
+title: Configuring calico/node
+---
+
+The `calico/node` container is primarily configured through environment variables.
+
+
+## Environment Variables
+
+| Environment | Description                              | Schema | Default |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: I'd like to see this table alphabetized.

File: master/reference/node/configuration.md [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+| IP6 | The IPv6 address for Calico will bind to. When specified, the address is saved in the  [node resource configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/node) for this host, overriding any previously configured value. When omitted, if an address has not yet been configured in the node resource, IPv6 routing is not enabled. When omitted, if an IPv6 address has been previously configured in the node resource, IPv6 is enabled using the already configured address. | ipv6 address | |
+| AS | The AS number for this node. When specified, the value is saved in the node resource configuration for this host, overriding any previously configured value. When omitted, if an AS number has been previously configured in the node resource, that AS number is used for the peering.  When omitted, if an AS number has not yet been configured in the node resource, the node will use the global value (managed through `calicoctl config set/get asnumber`). | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. | kubernetes, etcd | etcd |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: etcd is not a valid DATASTORE_TYPE - it's etcdv2

File: master/reference/node/configuration.md [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+| CALICO_LIBNETWORK_IFPREFIX | Interface prefix to use for the network interface within the Docker containers that have been networked by the Calico driver. | string | cali |
+| CALICO_NETWORKING_BACKEND | Describes which BGP networking backend to use| gobgp, bird, none | bird |
+| **etcd configuration** | | | |
+| ETCD_ENDPOINTS    | A comma separated list of etcd endpoints (optional, defaults to http://127.0.0.1:2379) | http://etcd1:2379 | |
+| ETCD_KEY_FILE     | Path to the etcd key file (optional)       | /etc/calico/key.pem | |
+| ETCD_CERT_FILE    | Path to the etcd client cert (optional)    | /etc/calico/cert.pem | |
+| ETCD_CA_CERT_FILE | Path to the etcd CA file (optional)        | /etc/calico/ca.pem | |
+
+The ETCD_CA_FILE, ETCD_CERT_FILE, and ETCD_KEY_FILE
+environment variables are required when using etcd with SSL/TLS.  The values
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: "The values provided here... template"  what template?  I think this whole paragraph should go.

File: master/reference/node/configuration.md [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+| AS | The AS number for this node. When specified, the value is saved in the node resource configuration for this host, overriding any previously configured value. When omitted, if an AS number has been previously configured in the node resource, that AS number is used for the peering.  When omitted, if an AS number has not yet been configured in the node resource, the node will use the global value (managed through `calicoctl config set/get asnumber`). | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. | kubernetes, etcd | etcd |
+| WAIT_FOR_DATASTORE | Wait for connection to datastore before starting. If a succesfull connection is not made, node will shutdown. | boolean | false |
+| CALICO_LIBNETWORK_ENABLED | Enables running the docker-libnetwork plugin directly in the calico/node container. | boolean | true |
+| CALICO_LIBNETWORK_IFPREFIX | Interface prefix to use for the network interface within the Docker containers that have been networked by the Calico driver. | string | cali |
+| CALICO_NETWORKING_BACKEND | Describes which BGP networking backend to use| gobgp, bird, none | bird |
+| **etcd configuration** | | | |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: We don't need a separate title for this.

File: master/reference/node/configuration.md [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+| AS | The AS number for this node. When specified, the value is saved in the node resource configuration for this host, overriding any previously configured value. When omitted, if an AS number has been previously configured in the node resource, that AS number is used for the peering.  When omitted, if an AS number has not yet been configured in the node resource, the node will use the global value (managed through `calicoctl config set/get asnumber`). | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. | kubernetes, etcd | etcd |
+| WAIT_FOR_DATASTORE | Wait for connection to datastore before starting. If a succesfull connection is not made, node will shutdown. | boolean | false |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: typo: succesfull

File: master/reference/node/configuration.md [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+| IP6 | The IPv6 address for Calico will bind to. When specified, the address is saved in the  [node resource configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/node) for this host, overriding any previously configured value. When omitted, if an address has not yet been configured in the node resource, IPv6 routing is not enabled. When omitted, if an IPv6 address has been previously configured in the node resource, IPv6 is enabled using the already configured address. | ipv6 address | |
+| AS | The AS number for this node. When specified, the value is saved in the node resource configuration for this host, overriding any previously configured value. When omitted, if an AS number has been previously configured in the node resource, that AS number is used for the peering.  When omitted, if an AS number has not yet been configured in the node resource, the node will use the global value (managed through `calicoctl config set/get asnumber`). | int | global AS number |
+| DATASTORE_TYPE | Type of datastore. | kubernetes, etcdv2 | etcd |
```
Comment (caseydavenport) [inline master/reference/node/configuration.md]: The default column still says "etcd"
